### PR TITLE
Introduce deviceOSProtobuf.getDefinition(name) to device-os-protobuf

### PR DESCRIPTION
**Pull request by joegoggins**
## Overview

Introduces main public API methods for this module:

- `encode(protobufMessageName, {data: "foo"})`
- `decode(protobufMessageName, buffer)`
- `getDefinition(protobufMessageName)`
- `getDefinitions()` / `getNamespaces()`

Story details: https://app.shortcut.com/particle/story/95530

This work unblocks https://app.shortcut.com/particle/story/95530/introduce-deviceosprotobuf-getdefinition-name-to-device-os-protobuf#activity-95768

which in turn unblocks: https://app.shortcut.com/particle/story/94987/add-usb-validate-fqc-commands-to-mfg-cli

Hoping we can merge this without too much tussle/yak-shaving so i can get back to the P2 mfg-cli motivation behind this work before too long.

## How to test?

1. `git checkout sc-95530/introduce-getDefinition-function`
2. `npm i`
3. `npm run test:ci`; observe linter, tests, and coverage stats.
4. Read the README.md; note anywhere things could be written more clearly.
5. Read the rendered `docs/reference.md`; ask questions if anything isn't clear.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		return rootObject[protobufMessageName];
+	}
+
+	static _getProtobufReplyMessage(protobufMessageName) {
+		return this._getProtobufMessage(protobufMessageName.replace('Request','Reply'));
+	}
+}
+
+const pbjsJSONString = fs.readFileSync(`${path.resolve(__dirname)}/pbjs-generated/index.json`);
```
Comment (monkbroc) [inline src/index.js:115]: Did the following not work?
```suggestion
const pbjsJSONString = require('./pbjs-generated/index.json`);
```
The `fs` module is not available in React Native

**Comment by joegoggins** [comment]
While critically reviewing this PR myself, I decided to read the entire README of the [protobufjs](https://www.npmjs.com/package/protobufjs) to understand if we are using this module correctly.

One aspect of our technical approach is odd.

We use both `npm build` / `pbjs -o static-module` (for generated js code) _and_ `npm build:json` / `pbjs -o json` (for the mapping to request id). This choice was made without considering the `protobuf.load("awesome.proto"...`/reflection option which can reflect on `*.proto` files at runtime rather than via pre-generated code. If this worked, that approach would eliminate these 2 build steps.

I'm not proposing we pursue that now because I think this PR is a good step forward. But wanted to raise the option in case others wanted to chime in.

File: src/index.test.js [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+			expect(msg).to.be.an.instanceOf(protobufDefinition.message);
+			const buffer = protobufDefinition.message.encode(msg).finish();
+			expect(buffer).to.be.an.instanceOf(Buffer);
+
+			// Nuance: We don't know how to stub the reply in these tests yet
+			// Normally, Device OS is responsible for creating the buffer representing
+			// the replyMsg.
+			// This code does not work because bufferFromDeviceOS is invalid
+			// It's unclear how to produce valid byutes to mock this for now
+			// const bufferFromDeviceOS = Buffer.alloc(15, '1');
```
Comment (monkbroc) [inline src/index.test.js:377]: In a terminal, you can load the correct Response object and encode a response, store it in hex in this file like this: `Buffer.from('e73434490154', 'hex')`

File: src/index.js (hunk 1 above)
Comment (monkbroc) [reply src/index.js:115]: I'd recommend renaming the file to `pbjs-generated/definitions.json` and not `index.json` so it's not confusing between `index.js` and `index.json`. There's a parameter for `pbjs` for the name of the output file, right?

**Review by monkbroc — APPROVED** [review]
Looks good at a glance. I haven't run the library directly.

**Comment by monkbroc** [comment]
> We use both `npm build` / `pbjs -o static-module` (for generated js code) _and_ `npm build:json` / `pbjs -o json`

I'd say run both variants of the `pbjs` command with one `npm run build`

File: src/index.js (hunk 1 above)
Comment (joegoggins) [reply src/index.js]: I did try eliminating `fs` and `path` starting with the code change you suggested above. At first, it looked like it didn't work. I just tried again and realized the problem was i was doing that `require` AND yet still running `JSON.parse` on it, aka double parsing. Fixed now.

File: src/index.test.js (hunk 2 above)
Comment (joegoggins) [reply src/index.test.js]: 'preciate the comment. Though the solution you alluded to didn't work for me, I did manage to achieve what I was hoping for via https://github.com/particle-iot/device-os-protobuf/pull/10/commits/881d054face1833ccf6310dd995f877d605df1c9 👍

**Comment by joegoggins** [comment]
Cool. Thanks for the feedback @monkbroc , I've pushed commits that address all of your comments/questions.

@sergeuz I'm looking forward to your feedback as well. 

One thing I forgot to mention earlier was this snippet of code inside `particle-usb` (see [experimental branch](https://github.com/particle-iot/particle-usb/pull/56)) that demonstrates how low level API vs high level API will operate with this `getDefinition()` construct from the new `sendProtobufRequest` method we'll introduce there. I'm into it, that PR review should be coming along shortly.

```
async function main() {
    const particleUSB = require('./src/particle-usb');
    const devices = await particleUSB.getDevices();
    const device = devices[0];
    await device.open();
    const serialNumberHighLevelAPI = await device.getSerialNumber();
    console.log('serialNumberHighLevelAPI', serialNumberHighLevelAPI);

    const serialNumberLowLevelAPI = await device.sendProtobufRequest(
        'GetSerialNumberRequest',
        null,
        {timeout: 7000}
    )
    console.log('serialNumberLowLevelAPI', serialNumberLowLevelAPI.serial);

    await device.close();
}

main();

```

File: package.json [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
   "name": "@particle/device-os-protobuf",
-  "version": "1.0.0",
+  "version": "1.1.0-2",
   "description": "Particle Device OS Protobuf Definitions",
-  "main": "dist/index.js",
+  "main": "src/index.js",
   "scripts": {
-    "build": "mkdirp dist && pbjs control/*.proto -t static-module -o dist/index.js --no-delimited --no-convert --no-verify",
-    "docs": "jsdoc -X dist/index.js > dist/index-docs.json",
+    "build": "npm run build:pbjs:module && npm run build:pbjs:json && npm run build:docs",
```
Comment (monkbroc) [inline package.json:7]: Nice :+1:

**Review by monkbroc — APPROVED** [review]
Looks great now! :shipit:

**Pull request merged**